### PR TITLE
Added Vec append to BufferVec - Issue #3531

### LINKED ISSUE #3531: Expose more of the Vec API on BufferVec
## What problem does this solve or what need does it fill?

More efficient mutation APIs for `BufferVec` than just `push`, `reserve`, and `clear`.  Repeatedly `push`ing large quantities can be quite inefficient due to needing to repeatedly do size and capacity checks on the underlying Vec.

## What solution would you like?

Expose more of the mutation APIs to more efficiently build `BufferVecs`. Most notably `append` for grouping chains of `Vec`s together. 

## What alternative(s) have you considered?

 - Repeatedly `push`ing the items onto the `BufferVec`.
 - `DerefMut` on the underlying `Vec`.

## Additional context

This would both simplify and speed up the "prepare" stage of rendering of entities with large quantities of mutable values, including but not limited to particle systems and large scale batching.

### PR DESCRIPTION
**Pull request by Connor-McMillin**
# Objective

- Fixes #3531 

## Solution

- Added an append wrapper to BufferVec based on the function signature for vec.append()

---

First PR to Bevy. I didn't see any tests for other BufferVec methods (could have missed them) and currently this method is not used anywhere in the project. Let me know if there are tests to add or if I should find somewhere to use append so it is not dead code. The issue mentions implementing `truncate` and `extend` which were already implemented and merged [here](https://github.com/bevyengine/bevy/pull/6833/files#diff-c8fb332382379e383f1811e30c31991b1e0feb38ca436c357971755368012ced)

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
I would like to see a couple of simple tests for this: seems like useful functionality both internally and extrenally but I agree that dead code is rough.

**Comment by Connor-McMillin** [comment]
Are there any other vector functions that should get added with this PR?

File: crates/bevy_render/src/render_resource/buffer_vec.rs [hunk 1]
```diff
@@ -75,6 +75,10 @@ impl<T: Pod> BufferVec<T> {
         index
     }
 
+    fn append(&mut self, other: &mut BufferVec<T>) {
```
Comment (mockersf) [inline crates/bevy_render/src/render_resource/buffer_vec.rs:78]: Could you make it public?

Comment (mockersf) [inline crates/bevy_render/src/render_resource/buffer_vec.rs:78]:
```suggestion
    pub fn append(&mut self, other: &mut BufferVec<T>) {
```
Could you make it public?

Comment (Connor-McMillin) [reply crates/bevy_render/src/render_resource/buffer_vec.rs]: Good catch! That also seems to have resolved the dead code error during build

**Comment by Connor-McMillin** [comment]
Thanks @mockersf for pointing out that I needed to make the function public. This also resolved the dead code warnings, although I don't know _why_ it doesn't detect this as dead code anymore.

The tests I wrote (and momentarily pushed to this branch)
- Didn't resolve the dead code warning
- Could not easily use a test struct that derives `Pod` because that requires a dependency on `bytemuck` which `bevy_render` doesn't require currently and requiring that dependency for one or two tests doesn't seem like a good idea
- Were pretty gross using Mat4 which was one of the objects I found that derives `Pod` and didn't require additional dependencies to other projects

So I think just the implementation sans tests probably makes the most sense and is in line with the rest of BufferVec not being tested directly

**Review by alice-i-cecile — APPROVED** [review]

**Comment by mockersf** [comment]
> Thanks @mockersf for pointing out that I needed to make the function public. This also resolved the dead code warnings, although I don't know why it doesn't detect this as dead code anymore.

Because it's now part of the exposed public API, so it can be used by caller of the lib

**Review by mockersf — APPROVED** [review]

**Review by james7132 — APPROVED** [review]

**Pull request merged**